cps/acl.c

```diff
@@ -0,0 +1,535 @@
+ */
+
+static void
+cps_arp_cb(const struct lls_map *map, void *arg, enum lls_reply_ty ty,
+	int *pcall_again)
+{
+	struct gatekeeper_if *iface = arg;
+	struct rte_mbuf *created_pkt;
+	struct ether_hdr *eth_hdr;
+	struct arp_hdr *arp_hdr;
+	size_t pkt_size;
+	struct cps_config *cps_conf = get_cps_conf();
+	struct rte_kni *kni;
+	struct rte_mempool *mp;
+	int ret;
+
+	RTE_VERIFY(ty == LLS_REPLY_RESOLUTION);
+	if (pcall_again != NULL)
+		*pcall_again = false;
+
+	mp = cps_conf->net->gatekeeper_pktmbuf_pool[
+		rte_lcore_to_socket_id(cps_conf->lcore_id)];
+	created_pkt = rte_pktmbuf_alloc(mp);
+	if (created_pkt == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: could not allocate an ARP reply on the %s KNI\n",
+			iface->name);
+		return;
+	}
+
+	pkt_size = sizeof(struct ether_hdr) + sizeof(struct arp_hdr);
+	created_pkt->data_len = pkt_size;
+	created_pkt->pkt_len = pkt_size;
+
+	/* Set-up Ethernet header. */
+	eth_hdr = rte_pktmbuf_mtod(created_pkt, struct ether_hdr *);
+	ether_addr_copy(&map->ha, &eth_hdr->s_addr);
+	ether_addr_copy(&iface->eth_addr, &eth_hdr->d_addr);
+	eth_hdr->ether_type = rte_cpu_to_be_16(ETHER_TYPE_ARP);
+
+	/* Set-up ARP header. */
+	arp_hdr = (struct arp_hdr *)(eth_hdr + 1);
+	arp_hdr->arp_hrd = rte_cpu_to_be_16(ARP_HRD_ETHER);
+	arp_hdr->arp_pro = rte_cpu_to_be_16(ETHER_TYPE_IPv4);
+	arp_hdr->arp_hln = ETHER_ADDR_LEN;
+	arp_hdr->arp_pln = sizeof(struct in_addr);
+	arp_hdr->arp_op = rte_cpu_to_be_16(ARP_OP_REPLY);
+	ether_addr_copy(&map->ha, &arp_hdr->arp_data.arp_sha);
+	rte_memcpy(&arp_hdr->arp_data.arp_sip, map->ip_be,
+		sizeof(arp_hdr->arp_data.arp_sip));
+	ether_addr_copy(&iface->eth_addr, &arp_hdr->arp_data.arp_tha);
+	arp_hdr->arp_data.arp_tip = iface->ip4_addr.s_addr;
+
+	if (iface == &cps_conf->net->front)
+		kni = cps_conf->front_kni;
+	else
+		kni = cps_conf->back_kni;
+
+	ret = rte_kni_tx_burst(kni, &created_pkt, 1);
+	if (ret <= 0) {
+		rte_pktmbuf_free(created_pkt);
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: could not transmit an ARP reply to the %s KNI\n",
+			iface->name);
+	}
+}
+
+void
+kni_acl_process_arp(struct cps_config *cps_conf, struct gatekeeper_if *iface,
+	struct rte_mbuf *buf, const struct ether_hdr *eth_hdr)
+{
+	struct arp_hdr *arp_hdr;
+	uint16_t pkt_len = rte_pktmbuf_data_len(buf);
+
+	/*
+	 * Just do basic length check, and assume that all other checks
+	 * are performed by the kernel before the packet is sent.
+	 */
+	if (unlikely(pkt_len < sizeof(*eth_hdr) + sizeof(*arp_hdr))) {
+		RTE_LOG(ERR, GATEKEEPER, "cps: KNI received ARP packet of size %hu bytes, but it should be at least %zu bytes\n",
+			pkt_len, sizeof(*eth_hdr) + sizeof(*arp_hdr));
+		goto out;
+	}
+
+	arp_hdr = rte_pktmbuf_mtod_offset(buf, struct arp_hdr *,
+		sizeof(*eth_hdr));
+
+	/* If it's a Gratuitous ARP or reply, then no action is needed. */
+	if (unlikely(rte_be_to_cpu_16(arp_hdr->arp_op) != ARP_OP_REQUEST ||
+			is_garp_pkt(arp_hdr)))
+		goto out;
+
+	hold_arp(cps_arp_cb, iface,
```

Comment:
BUG: consider that the kernel is trying to resolve IP address of a host that went offline. Each packet the kernel sends is going to register a new callback that will "never" call back.

We need a way to avoid registering duplicate callbacks, and callbacks must have a timeout.